File: windows/EMWaver.Tests/TransportDeviceConnectionStateTests.cs
using EMWaver.Services;
using Xunit;

namespace EMWaver.Tests;

public sealed class TransportDeviceConnectionStateTests
{
    [Fact]
    public void CurrentScriptDeviceIdFollowsActiveConnectionWhenPresent()
    {
        var state = new TransportDeviceConnectionState();
        var target = state.SetTarget(" usb:target ", DeviceTransport.UsbMidi);
        var connection = new FakeConnection("usb:connection", "USB Board");

        state.SetConnection(connection);

        Assert.Equal("usb:target", target.DeviceId);
        Assert.Equal("usb:connection", state.CurrentScriptDeviceId);
        Assert.Same(connection, state.Connection);
        Assert.True(state.MatchesDeviceId("USB:TARGET"));
        Assert.True(state.MatchesTransport(DeviceTransport.UsbMidi));
    }

    [Fact]
    public void ClearingMatchingTransportDropsConnectionAndResetsTarget()
    {
        var state = new TransportDeviceConnectionState();
        state.SetTarget("ble:board", DeviceTransport.Ble);
        state.SetConnection(new FakeConnection("ble:board", "BLE Board"));

        state.Clear(DeviceTransport.UsbMidi);
        Assert.Equal("ble:board", state.CurrentScriptDeviceId);

        state.Clear(DeviceTransport.Ble);
        Assert.Equal("active", state.CurrentScriptDeviceId);
        Assert.Equal(DeviceTransport.None, state.Transport);
        Assert.False(state.MatchesDeviceId("ble:board"));
    }

    [Fact]
    public void RetainsConnectionsByDeviceIdForFutureMultiDeviceRouting()
    {
        var state = new TransportDeviceConnectionState();
        var first = new FakeConnection("usb:board-1", "USB Board 1");
        var second = new FakeConnection("ble:board-2", "BLE Board 2");

        state.SetConnection(first);
        state.SetConnection(second);

        Assert.Same(first, state.ConnectionFor(" USB:BOARD-1 "));
        Assert.Same(second, state.ConnectionFor("ble:board-2"));
        Assert.Same(second, state.Connection);
    }

    private sealed class FakeConnection : ITransportDeviceConnection
    {
        internal FakeConnection(string sessionId, string displayName)
        {
            SessionId = sessionId;
            DisplayName = displayName;
            Session = new DeviceBufferSession(sessionId);
        }

        public string SessionId { get; }
        public string DisplayName { get; }
        public ITransportDeviceSession Session { get; }
    }
}

File: ios/EMWaverTests/TransportDeviceConnectionStateTests.swift
import XCTest
@testable import EMWaver

final class TransportDeviceConnectionStateTests: XCTestCase {
    private enum Transport {
        case none
        case usb
        case ble
    }

    func testCurrentScriptDeviceIdFollowsActiveConnectionWhenPresent() {
        let state = TransportDeviceConnectionState(noneTransport: Transport.none)
        let target = state.setTarget(deviceId: " usb:target ", transport: .usb)
        let connection = FakeConnection(sessionKey: "usb:connection", displayName: "USB Board")

        state.setConnection(connection)

        XCTAssertEqual(target.deviceId, "usb:target")
        XCTAssertEqual(state.currentScriptDeviceId, "usb:connection")
        XCTAssertEqual(state.connection?.sessionKey, connection.sessionKey)
        XCTAssertTrue(state.matchesDeviceId("USB:TARGET"))
        XCTAssertTrue(state.matchesTransport(.usb))
    }

    func testClearingMatchingTransportDropsConnectionAndResetsTarget() {
        let state = TransportDeviceConnectionState(noneTransport: Transport.none)
        state.setTarget(deviceId: "ble:board", transport: .ble)
        state.setConnection(FakeConnection(sessionKey: "ble:board", displayName: "BLE Board"))

        state.clear(transport: .usb)
        XCTAssertEqual(state.currentScriptDeviceId, "ble:board")

        state.clear(transport: .ble)
        XCTAssertEqual(state.currentScriptDeviceId, "active")
        XCTAssertEqual(state.transport, .none)
        XCTAssertFalse(state.matchesDeviceId("ble:board"))
    }

    private final class FakeConnection: TransportDeviceConnection {
        let sessionKey: String
        let displayName: String
        let session: TransportDeviceSession

        init(sessionKey: String, displayName: String) {
            self.sessionKey = sessionKey
            self.displayName = displayName
            self.session = DeviceBufferSession()
        }
    }
}
